src/Shared/FileUtilities.cs

```diff
@@ -53,8 +53,8 @@ internal static void ClearCacheDirectoryPath()
             cacheDirectory = null;
         }
 
-        // TODO: assumption on file system case sensitivity: https://github.com/Microsoft/msbuild/issues/781
-        internal static readonly StringComparison PathComparison = StringComparison.OrdinalIgnoreCase;
+        internal static readonly StringComparison PathComparison = Directory.Exists(Path.GetTempPath().ToUpper()) && Directory.Exists(Path.GetTempPath().ToLower()) ?
```

Comment:
Lurking here based on something that caught my eye, I have not reviewed the rest of the change.

Minor tweak:
```suggestion
        internal static readonly StringComparison PathComparison = Directory.Exists(Path.GetTempPath().ToUpperInvariant()) && Directory.Exists(Path.GetTempPath().ToLowerInvariant()) ?
```

I'm a also *little* worried about other edge cases here. For example, if both /tmp/ and /TMP/ exist separately on a case-sensitive file system, we'll take that to mean that we should use case-insensitive. 

Or what if TEMP/TMPDIR points to a location without characters that change with ToUpper/ToLower, say /_/?

There is also the small possibility that /tmp is on a case-sensitive file system, while the things we're comparing are not. 

I'm sure there are more. This is _hard_.

I don't know what would be better here, though. Maybe @JeremyKuhne has some advice... I vaguely recall some BCL features being discussed around this.

Reply:
What we do in the libraries is this: https://source.dot.net/#System.IO.FileSystem/Common/System/IO/PathInternal.CaseSensitivity.cs,41

I'd follow along with that and reference it to maintain consistency. As pointed out, however, it falls apart on systems with multiple mounted file systems. And those file systems can be mounted anywhere... The problem is likely unsolvable without a major shift in how we deal with paths that would likely not benefit many and cost everyone. :)

cc: @carlossanlop, fyi

Reply:
Is there anything tracking making that BCL decision publicly available?